include/swift/Frontend/FrontendOptions.h

```diff
@@ -110,21 +114,64 @@ class FrontendInputs {
 
   // Primary input readers
 
-  Optional<SelectedInput> getPrimaryInput() const { return PrimaryInput; }
-  bool hasPrimaryInput() const { return getPrimaryInput().hasValue(); }
+private:
+  void assertMustNotBeMoreThanOnePrimaryInput() const {
+    assert(PrimaryInputs.size() < 2 &&
+           "have not implemented >1 primary input yet");
+  }
+
+public:
+  ArrayRef<SelectedInput> getPrimaryInputs() const { return PrimaryInputs; }
+
+  unsigned primaryInputCount() const { return getPrimaryInputs().size(); }
+
+  // Primary count readers:
+
+  bool haveUniquePrimaryInput() const { return primaryInputCount() == 1; }
+
+  bool havePrimaryInputs() const { return primaryInputCount() > 0; }
 
-  bool isWholeModule() { return !hasPrimaryInput(); }
+  bool isWholeModule() const { return !havePrimaryInputs(); }
 
-  bool isPrimaryInputAFileAt(unsigned i) {
-    return hasPrimaryInput() && getPrimaryInput()->isFilename() &&
-           getPrimaryInput()->Index == i;
+  // Count-dependend readers:
+
+  Optional<SelectedInput> getOptionalPrimaryInput() const {
+    return havePrimaryInputs() ? Optional<SelectedInput>(getPrimaryInputs()[0])
+                               : Optional<SelectedInput>();
+  }
+
+  SelectedInput getRequiredUniquePrimaryInput() const {
+    assert(haveUniquePrimaryInput());
+    return getPrimaryInputs()[0];
+  }
+
+  Optional<SelectedInput> getOptionalUniquePrimaryInput() const {
+    return haveUniquePrimaryInput()
+               ? Optional<SelectedInput>(getPrimaryInputs()[0])
+               : Optional<SelectedInput>();
   }
+
   bool haveAPrimaryInputFile() const {
-    return hasPrimaryInput() && getPrimaryInput()->isFilename();
+    return havePrimaryInputs() && getOptionalPrimaryInput()->isFilename();
+  }
+
+  Optional<StringRef> getOptionalUniquePrimaryInputFilename() const {
+    Optional<SelectedInput> primaryInput = getOptionalUniquePrimaryInput();
+    return (primaryInput && primaryInput->isFilename())
+               ? Optional<StringRef>(getInputFilenames()[primaryInput->Index])
+               : Optional<StringRef>();
+  }
+
+  bool isPrimaryInputAFileAt(unsigned i) const {
```

Comment:
This was already present, but it's a weird name for the function, particularly because of the "a". It's not actually asking if the primary input is a (any?) file at index `i`; it's asking if there's a primary file and, if so, if it's at index `i`. If we were in a world with multiple primary files, it would be asking if the file at index `i` is a primary file.